sweepbatcher: add option WithCustomSignMuSig2

starius opened the pull request:
In this PR I'm adding option WithCustomSignMuSig2 which provides another way of producing musig2 signatures.

#### Pull Request Checklist
- [ ] Update `release_notes.md` if your PR contains major features, breaking changes or bugfixes

bhandras commented on sweepbatcher/sweep_batch.go:
```diff
@@ -715,10 +715,11 @@ func (b *batch) publishBatch(ctx context.Context) (btcutil.Amount, error) {
 
-	fee = b.rbfCache.FeeRate.FeeForWeight(weightEstimate.Weight())
+	weight := weightEstimate.Weight()
+	fee0 := b.rbfCache.FeeRate.FeeForWeight(weight)
```
nit: instead of `fee0` and `fee` we could call them `feeForWeight` and `clampedFee`.

starius replied:
I renamed to `feeForWeight` and `fee`.

bhandras commented on sweepbatcher/sweep_batch.go:
```diff
@@ -749,8 +750,9 @@ func (b *batch) publishBatch(ctx context.Context) (btcutil.Amount, error) {
 	b.log.Infof("attempting to publish non-coop tx=%v with feerate=%v, "+
-		"totalfee=%v, sweeps=%d, destAddr=%s", batchTx.TxHash(),
-		b.rbfCache.FeeRate, fee, len(batchTx.TxIn), address)
+		"weight=%v, fee0=%v, finalfee=%v, sweeps=%d, destAddr=%s",
```
nit: Similarly logging out `fee0` is not super informative, so perhaps we could as suggested above log something like: `calculatedfee` and `finalfee`

starius replied:
Here as well: `feeForWeight` and `fee`.

bhandras commented on sweepbatcher/sweep_batcher.go:
```diff
@@ -136,6 +137,12 @@ type MuSig2SignSweep func(ctx context.Context,
 	prevoutMap map[wire.OutPoint]*wire.TxOut) (
 	[]byte, []byte, error)
 
+// MuSig2SignSweep is a function that can be used to sign a sweep transaction
```
nit: godoc is off

starius replied:
Fixed

bhandras commented on sweepbatcher/sweep_batch.go:
```diff
@@ -998,6 +1004,36 @@ func (b *batch) musig2sign(ctx context.Context, inputIndex int, sweep sweep,
+	copy(digest[:], sigHash)
+
+	if b.cfg.signMuSig2 != nil {
+		// A custom signer is installed. Use it instead of
```
nit: i'd suggest to move the comment above the if for readability.

starius replied:
Fixed

bhandras commented on sweepbatcher/sweep_batch.go:
```diff
@@ -491,7 +497,7 @@ func (b *batch) Run(ctx context.Context) error {
 	}()
 
-	if b.muSig2SignSweep == nil {
+	if b.muSig2SignSweep == nil && b.cfg.signMuSig2 == nil {
```
Perhaps also add a check that only one is set.

bhandras commented on sweepbatcher/sweep_batcher.go:
```diff
@@ -265,6 +284,17 @@ func WithNoBumping() BatcherOption {
+// musig2 signatures instead of musig2SignSweep and signerClient. Note
+// that musig2SignSweep must be nil in this case, however signerClient
+// must still be provided, as it is used for non-coop spendings.
+func WithSignMuSig2(signMuSig2 SignMuSig2) BatcherOption {
```
nit: consider calling this `WithCustomMuSig2Signer`.

starius replied:
Fixed

bhandras commented on sweepbatcher/sweep_batcher_test.go:
```diff
@@ -1944,6 +1952,74 @@ func testSweepBatcherCloseDuringAdding(t *testing.T, store testStore,
+	// Use custom MuSig2 signer function.
+	batcher := NewBatcher(lnd.WalletKit, lnd.ChainNotifier, lnd.Signer,
+		nil, nil, lnd.ChainParams, batcherStore,
+		sweepStore, WithSignMuSig2(testSignMuSig2func))
```
I'd also check that the signature is as expected so we test that the custom signer was indeed used.

starius replied:
Fixed

bhandras reviewed (commented):
Looks good, just a few nits.

starius commented:
Rebased

starius commented:
I found that in many tests transactions were not successfully signed and broadcasted.
Fixed this and added checks that transactions are broadcasted (receiving from `lnd.TxPublishChannel` channel).

bhandras commented on sweepbatcher/sweep_batcher_test.go:
```diff
@@ -48,15 +47,22 @@ var destAddr = func() btcutil.Address {
 }()
 
-var senderKey, receiverKey [33]byte
+var htlcKeys = func() loopdb.HtlcKeys {
```
nit: instead of declaring as variable, you could just declare it as function.

starius replied:
Then key generation would happen every time this variables is used (many times in this file). I want to cache the values.

bhandras reviewed (approved):
LGTM 🚀

hieblmi commented on sweepbatcher/sweep_batch.go:
```diff
@@ -942,138 +940,154 @@ func (b *batch) debugLogTx(msg string, tx *wire.MsgTx) {
+	// Now we're creating a local MuSig2 session using the receiver
+	// key's key locator and the htlc's root hash.
+	musig2SessionInfo, err := b.signerClient.MuSig2CreateSession(
+		ctx, muSig2Version,
```
nit: early line break

starius replied:
Fixed

hieblmi commented on sweepbatcher/sweep_batch.go:
```diff
@@ -942,138 +940,154 @@ func (b *batch) debugLogTx(msg string, tx *wire.MsgTx) {
+	musig2SessionInfo, err := b.signerClient.MuSig2CreateSession(
+		ctx, muSig2Version,
+		&sweep.htlcKeys.ClientScriptKeyLocator, signers,
+		lndclient.MuSig2TaprootTweakOpt(
```
nit: fits in one line

starius replied:
Fixed

hieblmi commented on sweepbatcher/sweep_batch.go:
```diff
@@ -942,138 +940,154 @@ func (b *batch) debugLogTx(msg string, tx *wire.MsgTx) {
+	// To be sure that we're good, parse and validate that the
+	// combined signature is indeed valid for the sig hash and the
+	// internal pubkey.
+	err = b.verifySchnorrSig(
```
nit: fits in one line

starius replied:
Fixed

hieblmi commented on sweepbatcher/sweep_batch.go:
```diff
@@ -942,138 +940,154 @@ func (b *batch) debugLogTx(msg string, tx *wire.MsgTx) {
-		sigHash, err := txscript.CalcTaprootSignatureHash(
-			sigHashes, txscript.SigHashDefault, packet.UnsignedTx,
-			i, prevOutputFetcher,
+		finalSig, err := b.musig2sign(
```
👍

hieblmi commented on sweepbatcher/sweep_batch.go:
```diff
@@ -137,6 +137,12 @@ type batchConfig struct {
+	// musig2 signatures instead of musig2SignSweep and signerClient. Note
+	// that musig2SignSweep must be nil in this case, however signerClient
+	// must still be provided, as it is used for non-coop spendings.
+	signMuSig2 SignMuSig2
```
can we rename this to `customMuSig2Signer`? Lines like `if cfg.signMuSig2 != nil && musig2ServerSigner != nil` then become easier to read.

starius replied:
Fixed

hieblmi commented on sweepbatcher/sweep_batcher.go:
```diff
@@ -245,6 +252,12 @@ type Batcher struct {
+	// musig2 signatures instead of musig2SignSweep and signerClient. Note
+	// that musig2SignSweep must be nil in this case, however signerClient
+	// must still be provided, as it is used for non-coop spendings.
+	signMuSig2 SignMuSig2
```
same here, better name it something explicit like `customMuSig2Signer`.

starius replied:
Fixed

hieblmi commented on sweepbatcher/sweep_batcher.go:
```diff
@@ -254,6 +267,12 @@ type BatcherConfig struct {
+	// musig2 signatures instead of musig2SignSweep and signerClient. Note
+	// that musig2SignSweep must be nil in this case, however signerClient
+	// must still be provided, as it is used for non-coop spendings.
+	signMuSig2 SignMuSig2
```
same here, better name it something explicit like `customMuSig2Signer`.

starius replied:
Fixed

hieblmi reviewed (approved):
The changes LGTM, thank you for the extension! Let's maybe rename the signer to something more concise.

starius commented on sweepbatcher/sweep_batch.go:
```diff
@@ -942,138 +940,154 @@ func (b *batch) debugLogTx(msg string, tx *wire.MsgTx) {
 
-		var serverPublicNonce [musig2.PubNonceSize]byte
-		copy(serverPublicNonce[:], serverNonce)
+	sigHashes := txscript.NewTxSigHashes(
```
Thanks! Fixed.

starius commented on sweepbatcher/sweep_batch.go:
```diff
@@ -942,138 +940,154 @@ func (b *batch) debugLogTx(msg string, tx *wire.MsgTx) {
-		if err != nil {
-			return err
-		}
+	// Now we're creating a local MuSig2 session using the receiver
```
Fixed

starius commented on sweepbatcher/sweep_batch.go:
```diff
@@ -942,138 +940,154 @@ func (b *batch) debugLogTx(msg string, tx *wire.MsgTx) {
-		}
+	// Now we're creating a local MuSig2 session using the receiver
+	// key's key locator and the htlc's root hash.
+	musig2SessionInfo, err := b.signerClient.MuSig2CreateSession(
```
Thanks! Done
